solana/ts/lib/quoter.ts

```diff
@@ -0,0 +1,432 @@
+  Connection,
+  LAMPORTS_PER_SOL,
+  PublicKey,
+  PublicKeyInitData,
+  SystemProgram,
+} from "@solana/web3.js";
+import { amount, chainToPlatform } from "@wormhole-foundation/sdk";
+import {
+  Chain,
+  deserializeLayout,
+  toChainId,
+} from "@wormhole-foundation/sdk-base";
+import IDL from "../idl/2_0_0/json/ntt_quoter.json";
+import { NttQuoter as Idl } from "../idl/2_0_0/ts/ntt_quoter.js";
+import {
+  U64,
+  chainToBytes,
+  derivePda,
+  programDataAddress,
+  programDataLayout,
+  quoterAddresses,
+} from "./utils.js";
+
+//constants that must match ntt-quoter lib.rs / implementation:
+const USD_UNIT = 1e6;
+const GWEI_PER_ETH = 1e9;
+export const WEI_PER_GWEI = 1e9;
+
+const SEED_PREFIX_INSTANCE = "instance";
+const SEED_PREFIX_REGISTERED_CHAIN = "registered_chain";
+const SEED_PREFIX_REGISTERED_NTT = "registered_ntt";
+const SEED_PREFIX_RELAY_REQUEST = "relay_request";
+
+export class NttQuoter {
+  readonly instance: PublicKey;
+  private readonly program: Program<Idl>;
+  nttProgramId: PublicKey;
+  pdas: ReturnType<typeof quoterAddresses>;
+
+  constructor(
+    connection: Connection,
+    programId: PublicKeyInitData,
+    nttProgramId: PublicKeyInitData
+  ) {
+    // @ts-ignore
+    this.program = new Program<Idl>(IDL as Idl, new PublicKey(programId), {
+      connection,
+    });
+    this.nttProgramId = new PublicKey(nttProgramId);
+    this.pdas = quoterAddresses(this.program.programId);
+    this.instance = derivePda([SEED_PREFIX_INSTANCE], this.program.programId);
+  }
+
+  // ---- user relevant functions ----
+
+  async isRelayEnabled(destination: Chain) {
+    try {
+      const { paused } = await this.getRegisteredChain(destination);
+      return !paused;
+    } catch (e: any) {
+      if (e.message?.includes("Account does not exist")) {
+        return false;
+      }
+      throw e;
+    }
+  }
+
+  async calcRelayCostInSol(
+    nttProgramId: PublicKey,
+    chain: Chain,
+    requestedGasDropoffEth: number
+  ) {
+    const [chainData, nttData, instanceData, rentCost] = await Promise.all([
+      this.getRegisteredChain(chain),
+      this.getRegisteredNtt(nttProgramId),
+      this.getInstance(),
+      this.program.provider.connection.getMinimumBalanceForRentExemption(
+        this.program.account.relayRequest.size
+      ),
+    ]);
+
+    if (requestedGasDropoffEth > chainData.maxGasDropoffEth)
+      throw new Error("Requested gas dropoff exceeds allowed maximum");
+
+    const totalNativeGasCostUsd =
+      chainData.nativePriceUsd *
+      (requestedGasDropoffEth +
+        (chainData.gasPriceGwei * nttData.gasCost) / GWEI_PER_ETH);
+
+    const totalCostSol =
+      rentCost / LAMPORTS_PER_SOL +
+      (chainData.basePriceUsd + totalNativeGasCostUsd) /
+        instanceData.solPriceUsd;
+
+    return totalCostSol;
+  }
+
+  /**
+   * Estimate the cost of a relay request
+   * @param chain The destination chain
+   * @param gasDropoff The amount of native gas to end up with on the destination chain
```

Comment:
Does not state units. Using atomic instead of a human amount is inconsistent with the rest of the class.

I'd write a clean implementation and add this one with a preface that it is a required interface for an integration that wouldn't exist in a perfect world.